// @spreeloop/orange_money/dist/src/disbursements/utils/status.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DisbursementApiRawStatus = exports.DisbursementStatus = void 0;
/**
 * The simplified version of transaction status.
 */
var DisbursementStatus;
(function (DisbursementStatus) {
    /**
     * The transaction failed.
     */
    DisbursementStatus["failed"] = "FAILED";
    /**
     * The transaction is pending.
     */
    DisbursementStatus["pending"] = "PENDING";
    /**
     * The transaction succeeded.
     */
    DisbursementStatus["succeeded"] = "SUCCEEDED";
    /**
     * The status received from the endpoint was not recognized.
     */
    DisbursementStatus["unknown"] = "UNKNOWN";
})(DisbursementStatus = exports.DisbursementStatus || (exports.DisbursementStatus = {}));
var DisbursementApiRawStatus;
(function (DisbursementApiRawStatus) {
    /**
     * Transaction is in progress on Orange system.
     */
    DisbursementApiRawStatus["pending"] = "PENDING";
    /**
     * The user canceled the disbursement.
     */
    DisbursementApiRawStatus["canceled"] = "CANCELLED";
    /**
     * Waiting for user entry.
     */
    DisbursementApiRawStatus["initialized"] = "INITIATED";
    /**
     * Disbursement is done for mobile.
     */
    DisbursementApiRawStatus["succeeded"] = "SUCCESSFULL";
    /**
     * Disbursement is done for web.
     */
    DisbursementApiRawStatus["succeeded2"] = "SUCCESS";
    /**
     * Disbursement failed.
     */
    DisbursementApiRawStatus["failed"] = "FAILED";
    /**
     * The token timed out.
     * Note that the minimum token expiration time is 7 min.
     */
    DisbursementApiRawStatus["expired"] = "EXPIRED";
})(DisbursementApiRawStatus = exports.DisbursementApiRawStatus || (exports.DisbursementApiRawStatus = {}));
